### ISSUE TITLE
bulk_update Crashes When Updating NULLs

### ISSUE THREAD

**Issue by NinjaMeTimbers** [coleifer/peewee#2865]
I've run into an issue when trying to use bulk_update to update data in a postgresql database where if the only records for a non-text field being updated are NULL, the query fails because postgres doesn't know how to cast the NULL value and automatically assumes it's text.

Suppose for example the following model (simplified from what was generating the issue:

```
class Timeseries(BaseModel):
    locationid = ForeignKeyField(column_name='locationid', field='id', model=TblLocation)
    date = DateTimeField(index=True)
    value = FloatField(null=True)
    
    class Meta:
        table_name = 'tbl_timeseries'
        schema = schemaName
```

Now suppose I am trying to update value = NULL on a single record with id = 44629816.

Using Timeseries.bulk_update() would generate the following SQL (simplified for the example):

```
UPDATE "ww"."tbl_timeseries"
SET "locationid" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN 49770 END,
	"date" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN '2023-02-04 00:00:00' END,
	"value" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN NULL END
WHERE ("tbl_timeseries"."id" IN (44629816))
```

Running this SQL gives the following error:

```
SQL Error [42804]: ERROR: column "value" is of type double precision but expression is of type text
  Hint: You will need to rewrite or cast the expression.
```

From my investigating, the reason is that there's nothing in the CASE statement that tells postgres what type the value is supposed to be, and for some reason it can't just automatically cast NULL from a CASE statement to the datatype of the field ("value" = NULL still works fine though, it's only if it's wrapped in a case statement). Instead it just naively assumes that the NULL is a TEXT value, which it can't update a FLOAT value to. I did some testing and found that it also works if any value in the CASE statement has a discernible type. For example:

```
UPDATE "ww"."tbl_timeseries"
SET 	"value" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN NULL WHEN 12345 THEN 5 else null END
WHERE ("tbl_timeseries"."id" IN (44629816, 12345))
```

would work, but if the second value was NULL instead of 5, it would crash.

When all results of the CASE statement evaluate to NULL, it needs to be cast to the correct data type. For example, this solves the issue:

```
UPDATE "ww"."tbl_timeseries"
SET "locationid" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN 49770 END,
	"date" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN '2023-02-04 00:00:00' END,
	"value" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN NULL END::float
WHERE ("tbl_timeseries"."id" IN (44629816))
```

And likewise this would also solve the issue:

```
UPDATE "ww"."tbl_timeseries"
SET "locationid" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN 49770 END,
	"date" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN '2023-02-04 00:00:00' END,
	"value" = CASE "tbl_timeseries"."id" WHEN 44629816 THEN NULL::float END
WHERE ("tbl_timeseries"."id" IN (44629816))
```

So basically to handle NULLs, update_many needs to cast the values to the appropriate data type, or else the method crashes.

**Comment by coleifer** [comment]
Yeah, postgres has this issue with data-types. See also: https://github.com/coleifer/peewee/issues/2308

I suggest:

1. write your case statement manually
2. use regular updates instead of the big CASE

If you do (2) it'll likely be just as fast if done inside a transaction. As usual, profile to verify.
